### DISCUSSION TITLE
What is the reason for `ObjectMapper.writeValueAsString(Object value)` to throw a `JsonProcessingException`

### DISCUSSION THREAD

**Discussion by gusega** [FasterXML/jackson-databind#4370] [Q&A]
### Describe your Issue

Hi

What is the reason for `ObjectMapper.writeValueAsString(Object value)` to declare that it throws a `JsonProcessingException`. The method description says that this method can serialise _any_ java object to String.  So it should not and cannot throw any exceptions, or why does it?

The reason I am asking is because when I use this method, I must provide try-catch clause for the `JsonProcessingException` and I really do not understand, why I am doing it or how this could happen, so I end up adding a comment /*must never happen*/ and rethrow it as a runtime exception. If there is a legitimate reason or possibility for the method to throw `JsonProcessingException` could you please provide one? Then I could at least have some meaningful logic to deal with it.

I looked at the source code of `ObjectMapper.writeValueAsString()` and I don't see how or why it would throw a `JsonProcessingException`. The methods it calls throw `IOException` which makes it hard to pinpoint where and if `JsonProcessingException` exception is thrown if it's thrown at all, but it's a separate problem.

I looked at libraries that use this method and some of them call it in try - catch with empty catch clause, basically ignoring it. Others re-throw it as some kind of runtime exception.

Please advise.

Thanks.

**Comment by cowtowncoder** [comment]
The reason is because interfaces this method has to use declare throwing `IOException` and subtypes (`JsonProcessingException` and so on).
But since we know for a fact that no I/O operation should be used (since output target is `StringWriter`, not `Writer` or `OutputStream`), code does add bogus catch clause to only allow Jackson exception that may be thrown -- even if `IOException`s cannot. For example, value given might be non-serializable type, or have annotations that are incorrectly specified, leading to exception.

So despite method being theoretically usable on any given type, there are error conditions that may mean that in actuality some types (or possibly instances) will fail to be serialized. Such condition is reported as one of `JsonProcessingException` subtypes.

I would not recommend swallowing exception, except if method can signal error condition in some other way (like returning `null` which `writeValueAsString()` will never return; or `Optional.empty()`), unless specific type is known not to ever fail (like it being one of well-known JDK types, or `JsonNode`)
